clientconn.go

```diff
@@ -1159,21 +1167,25 @@ func (ac *addrConn) getReadyTransport() (transport.ClientTransport, bool) {
 func (ac *addrConn) tearDown(err error) {
 	ac.cancel()
 	ac.mu.Lock()
-	defer ac.mu.Unlock()
 	if ac.state == connectivity.Shutdown {
+		ac.mu.Unlock()
 		return
 	}
+	// We have to set the state to Shutdown before we call ac.transports.GracefulClose, because it signals to
+	// onClose not to try reconnecting the transport.
+	ac.state = connectivity.Shutdown
+	ac.tearDownErr = err
+	ac.cc.handleSubConnStateChange(ac.acbw, ac.state)
 	ac.curAddr = resolver.Address{}
 	if err == errConnDrain && ac.transport != nil {
 		// GracefulClose(...) may be executed multiple times when
 		// i) receiving multiple GoAway frames from the server; or
 		// ii) there are concurrent name resolver/Balancer triggered
 		// address removal and GoAway.
+		ac.mu.Unlock()
```

Comment:
Why do we release the lock around `GracefulClose` now?  (Comment?)

Reply:
Done (tldr: gracefulclose => close => onclose locks ac.mu)